Update README (#137)
trinodb/trino-python-client: README.md

> Trino client

Reviewer: nit: Trino Python client

Author: DONE. I think we need to change title of [trino-go-client](https://github.com/trinodb/trino-go-client) as well

[![Build Status](https://github.com/trinodb/trino-python-client/workflows/ci/badge.svg)](https://github.com/trinodb/trino-python-client/actions?query=workflow%3Aci+event%3Apush+branch%3Amaster)
[![Trino Slack](https://img.shields.io/static/v1?logo=slack&logoColor=959DA5&label=Slack&labelColor=333a41&message=join%20conversation&color=3AC358)](https://trino.io/slack.html)
[![Trino: The Definitive Guide book download](https://img.shields.io/badge/Trino%3A%20The%20Definitive%20Guide-download-brightgreen)](https://www.starburst.io/info/oreilly-trino-guide/)

## Usage
### The Python Database API (DBAPI)

**Installation**
```
$ pip install trino
```

**Quick Start**

Use the DBAPI interface to query Trino:

```python
from trino.dbapi import connect

conn = connect(
    host="<host>",
    port=<port>,
    user="<username>",
    catalog="<catalog>",
    schema="<schema>",
)
cur = conn.cursor()
cur.execute("SELECT * FROM system.runtime.nodes")
rows = cur.fetchall()
```

This will query the `system.runtime.nodes` system tables that shows the nodes in the Trino cluster.

The DBAPI implementation in `trino.dbapi` provides methods to retrieve fewer rows

> for example `Cursorfetchone()` or `Cursor.fetchmany()`. By default `Cursor.fetchmany()` fetches one row.

Reviewer: `Cursor.fetchone()`

Author: fixed

Author: P/S: I just check the old docs and the code, I think `Cursor.fetchmany()` is correct.
https://github.com/trinodb/trino-python-client/blob/62f53c9d3ed9978e8a2bcd0eeacf2c7f6e071959/trino/dbapi.py#L481-L482

        ...
    )
    ```

* SQLAlchemy
    ```python
    from sqlalchemy import create_engine
    from trino.auth import OAuth2Authentication

    engine = create_engine(
    "trino://<username>@<host>:<port>/<catalog>",
        connect_args={
            "http_scheme": "https",
            "auth": OAuth2Authentication(),
        }
    )
    ```

### Kerberos Authentication
The `KerberosAuthentication` class can be used to connect to a Trino cluster configured with
the [`Kerberos` authentication type](https://trino.io/docs/current/security/kerberos.html):

* DBAPI
    ```python
    from trino.dbapi import connect
    from trino.auth import KerberosAuthentication
    
    conn = connect(
        http_scheme="https",
        user="<username>"
        auth=KerberosAuthentication(...),
        ...
    )
    ```

* SQLAlchemy
    ```python
    from sqlalchemy import create_engine
    from trino.auth import KerberosAuthentication

    engine = create_engine(
        "trino://<username>@<host>:<port>/<catalog>",
        connect_args={
            "http_scheme": "https",
            "auth": KerberosAuthentication(...),
        }
    )
    ```

### User impersonation

> In the case of user who submit the query is not the same as user who authenticate to Trino server (e.g in SuperSet),

Reviewer: Superset

Trino Python client
===================
A [Trino](https://trino.io/) client for the [Python](https://www.python.org/) programming language.

[![Build Status](https://github.com/trinodb/trino-python-client/workflows/ci/badge.svg)](https://github.com/trinodb/trino-python-client/actions?query=workflow%3Aci+event%3Apush+branch%3Amaster)
[![Trino Slack](https://img.shields.io/static/v1?logo=slack&logoColor=959DA5&label=Slack&labelColor=333a41&message=join%20conversation&color=3AC358)](https://trino.io/slack.html)
[![Trino: The Definitive Guide book download](https://img.shields.io/badge/Trino%3A%20The%20Definitive%20Guide-download-brightgreen)](https://www.starburst.io/info/oreilly-trino-guide/)

## Usage
### The Python Database API (DBAPI)

**Installation**
```
$ pip install trino
```

**Quick Start**

Use the DBAPI interface to query Trino:

```python
from trino.dbapi import connect

conn = connect(
    host="<host>",
    port=<port>,
    user="<username>",
    catalog="<catalog>",
    schema="<schema>",
)
cur = conn.cursor()
cur.execute("SELECT * FROM system.runtime.nodes")
rows = cur.fetchall()
```

> This will query the `system.runtime.nodes` system tables that shows the nodes in the Trino cluster.

Reviewer 2: I would recommend extracting a commit for column width changes.

Author: I don't think it change the column width because there are some lines event longer, e.g
https://github.com/trinodb/trino-python-client/blob/65506e8f1c90ea8908576d43037be46f53c8604a/README.md?plain=1#L199

Reviewer 2: I suggested separating into two commits. 
* Add SQLArchemy example in README
* Change column width in README

Author: @ebyhr the column width config for now is `120`, right?
Let me re-format the code to match this rule.

Author: @ebyhr For line contain link, it's quite long (for example this one is 162 chars) but can't be break in to 2 lines.
Is it ok to keep it in one line? 
![image](https://user-images.githubusercontent.com/6848311/153702956-61bb9f70-a574-4b94-ab47-25675db0744a.png)

Reviewer 2: The 120 is for Python files, not markdown. 

> Is it ok to keep it in one line?

Yes, it's ok.

Author: @ebyhr yes, I known. But I don't find any docs/config talking about markdown convention.

Author: Resolved as personal discuss.

connection = engine.connect()

rows = connection.execute(text("SELECT * FROM runtime.nodes")).fetchall()

# or using SQLAlchemy schema
nodes = Table(
    'nodes',
    MetaData(schema='runtime'),
    autoload=True,
    autoload_with=engine
)
rows = connection.execute(select(nodes)).fetchall()
```

## Authentications
### Basic Authentication
The `BasicAuthentication` class can be used to connect to a Trino cluster configured with
the [Password file authentication type, LDAP authentication type or Salesforce authentication type](https://trino.io/docs/current/security/authentication-types.html):

* DBAPI
    ```python
    from trino.dbapi import connect
    from trino.auth import BasicAuthentication

    conn = connect(
        http_scheme="https",
        user="<username>"
        auth=BasicAuthentication("<username>", "<password>"),
        ...
    )
    ```

* SQLAlchemy
    ```python
    from sqlalchemy import create_engine
    
    engine = create_engine("trino://<username>:<password>@<host>:<port>/<catalog>")
    
    # or
    from trino.auth import BasicAuthentication
    engine = create_engine(
        "trino://<username>@<host>:<port>/<catalog>",
        connect_args={

>             http_scheme="https",

Reviewer 2: ```suggestion
            "http_scheme": "https",
```

            "auth": BasicAuthentication("<username>", "<password>"),
        }
    )
    ```

### JWT Authentication
The `JWTAuthentication` class can be used to connect to a Trino cluster configured with
the [`JWT` authentication type](https://trino.io/docs/current/security/jwt.html):

* DBAPI
    ```python
    from trino.dbapi import connect
    from trino.auth import JWTAuthentication
    
    conn = connect(
        http_scheme="https",
        user="<username>"
        auth=JWTAuthentication("<jwt_token>"),
        ...
    )
    ```

* SQLAlchemy
    ```python
    from sqlalchemy import create_engine
    from trino.auth import JWTAuthentication
    
    engine = create_engine(
        "trino://<username>@<host>:<port>/<catalog>",
        connect_args={

>             "auth": JWTAuthentication("<jwt_token>"),

Reviewer 2: `"http_scheme": "https",` looks missing?

rows = cur.fetchall()
```

This will query the `system.runtime.nodes` system tables that shows the nodes in the Trino cluster.

The DBAPI implementation in `trino.dbapi` provides methods to retrieve fewer
rows for example `Cursorfetchone()` or `Cursor.fetchmany()`. By default
`Cursor.fetchmany()` fetches one row. Please set
`trino.dbapi.Cursor.arraysize` accordingly.

### 1.3 Authentications

#### Basic Authentication

The `BasicAuthentication` class can be used to connect to a Trino cluster configured with
the [Password file authentication type, LDAP authentication type or Salesforce authentication type](https://trino.io/docs/current/security/authentication-types.html):

```python
from trino.dbapi import connect
from trino.auth import BasicAuthentication

conn = connect(
    http_scheme='https',

>     auth=BasicAuthentication("principal id", "password"),

Reviewer 3: Revert.

The principal is a different thing from the `user` that logs in.

Author: Yeah, I know.
Add a note to explain about `username` & `principal_id`

Author: Move to `User impersonation` session

  the redirect url to stdout

```python
from trino.dbapi import connect
from trino.auth import OAuth2Authentication

conn = connect(
    http_scheme='https',
    auth=OAuth2Authentication(),
    ...
)
```

### 1.4 Transactions

The client runs by default in *autocommit* mode. To enable transactions, set
*isolation_level* to a value different than `IsolationLevel.AUTOCOMMIT`:

```python
from trino.dbapi import connect
from trino.transaction import IsolationLevel

with connect(
        isolation_level=IsolationLevel.REPEATABLE_READ,
        ...
) as conn:
    cur = conn.cursor()
    cur.execute('INSERT INTO sometable VALUES (1, 2, 3)')
    cur.fetchall()
    cur.execute('INSERT INTO sometable VALUES (4, 5, 6)')
    cur.fetchall()
```

The transaction is created when the first SQL statement is executed.
`trino.dbapi.Connection.commit()` will be automatically called when the code exits the *with* context and the queries
succeed, otherwise
`trino.dbapi.Connection.rollback()` will be called.

## 2. SQLAlchemy

### 2.1 Installation

```
$ pip install trino[sqlalchemy]
```

### 2.2 Usage

> To connect from SQLAlchemy to Trino, use connection string (URL) following this pattern:

Reviewer 3: ```suggestion
To connect to Trino using SQLAlchemy, use a connection string (URL) following this pattern:
```

```

> trino://<username>:<password>@<host>:<port>/catalog/[schema]

Reviewer 3: ```suggestion
trino://<username>:<password>@<host>:<port>/<catalog>/<schema>
```

Also mention that the schema and password are optional (not everyone might be familiar that `[ ]` represents optional components).

```

Examples:

```python
from sqlalchemy import create_engine
from sqlalchemy.schema import Table, MetaData

> from sqlalchemy.sql.expression import select, text

Reviewer 3: Use fully-qualified names instead of importing symbols, it's helpful for new people since it's obvious where something is coming from.

engine = create_engine('trino://user@localhost:8080/system')
connection = engine.connect()

nodes = Table(
    'nodes',
    MetaData(schema='runtime'),
    autoload=True,
    autoload_with=engine
)

> connection.execute(select([nodes])).fetchall()
> 
> connection.execute(text("SELECT * FROM runtime.nodes")).fetchall()

Reviewer 3: SQLAlchemy no longer recommends passing tables as list (in preparation for 2.x style API).

```suggestion
rows = connection.execute(select([nodes])).fetchall()
```

If you  want to introduce people to SQLAlchemy's direct query mechanism, add a complete example like (but I'd suggest removing it):

```python
import sqlalchemy

engine = sqlalchemy.create_engine('trino://user@localhost:8080/system')
connection = engine.connect()
rows = connection.execute(sqlalchemy.sql.expression.text("SELECT * FROM runtime.nodes")).fetchall()
```

`Cursor.fetchmany()` fetches one row. Please set
`trino.dbapi.Cursor.arraysize` accordingly.

### SQLAlchemy

**Prerequisite**

- SQLAlchemy >= 1.3

> - Trino server >= 352

Author: Yeah. It's done.

        ...
    )
    ```

* SQLAlchemy

    ```python
    from sqlalchemy import create_engine
    from trino.auth import OAuth2Authentication

    engine = create_engine(
    "trino://<username>@<host>:<port>/<catalog>",
        connect_args={
            "auth": OAuth2Authentication(),
            "http_scheme": "https",
        }
    )
    ```

### Certificate Authentication

`CertificateAuthentication` class can be used to connect to Trino cluster configured with [certificate based authentication](https://trino.io/docs/current/security/certificate.html). `CertificateAuthentication` requires paths to a valid client certificate and private key.

- DBAPI

    ```python
    from trino.dbapi import connect
    from trino.auth import CertificateAuthentication

    conn = connect(
        user="<username>",
        auth=CertificateAuthentication("/path/to/cert.pem", "/path/to/key.pem"),
        http_scheme="https",
        ...
    )
    ```

- SQLAlchemy

    ```python
    from sqlalchemy import create_engine
    from trino.auth import CertificateAuthentication

    engine = create_engine(
    "trino://<username>@<host>:<port>/<catalog>",
        connect_args={

>             auth=CertificateAuthentication("/path/to/cert.pem", "/path/to/key.pem"),

Reviewer 2: ```suggestion
            "auth": CertificateAuthentication("/path/to/cert.pem", "/path/to/key.pem"),
```

Author: Oh, my mistake. Thanks
